src/main/java/net/oauth/jsontoken/JsonTokenParser.java

```diff
@@ -103,12 +105,18 @@ public JsonToken verifyAndDeserialize(String tokenString) throws SignatureExcept
    */
   private List<Verifier> provideVerifiers(JsonToken jsonToken) {
     Preconditions.checkNotNull(verifierProviders);
-    List<Verifier> verifiers = verifierProviders
-        .getVerifierProvider(jsonToken.getSignatureAlgorithm())
-        .findVerifier(jsonToken.getIssuer(), jsonToken.getKeyId());
+    SignatureAlgorithm signatureAlgorithm = jsonToken.getSignatureAlgorithm();
+    VerifierProvider provider = verifierProviders.getVerifierProvider(signatureAlgorithm);
+    if (provider == null) {
+      throw new IllegalArgumentException("Signature algorithm not supported: "
+          + signatureAlgorithm);
+    }
+
+    List<Verifier> verifiers = provider.findVerifier(jsonToken.getIssuer(), jsonToken.getKeyId());
```

Comment:
It is a bit strange for a function with List<> return type to return null. Basically, we need the caller to be aware of the distinction between null and empty list return value. In the VerifierProvider, what's the difference of null and empty list? Is it possible to return empty list in the Null cases?

Reply:
According to the documentation, VerifierProvider would return a list of possible verifier keys if keyId mismatches. So, if the VerifierProvider returns an empty list, it could be the case that the issuer exists in the system, but they just don't have any keys in the system (keys haven't yet been added, keys expired, keys revoked, etc). The documentation never explicitly mentions it, but I imagine returning null would imply that the issuer doesn't exist in the system at all, so there are no verifiers possible. In certain cases, it can be useful to distinguish between these two scenarios (different errors, logging, etc).

Reply:
I don't think the library would need to distinguish between those cases because it doesn't care whether the verifier is missing because of the issuer not existing or the key not existing. That should be up to the caller and both cases would be covered by the same NO_VERIFIER error. As Longsheng mentioned it's usually a code smell to have different handling between empty list and null list, and if the library cared to distinguish those cases it would probably be better to have another more explicit way to distinguish those cases rather than assuming it based on empty vs null response.

However, the VerifierProvider interface is implemented by callers of the library and we can't prevent them from returning null here now because that might be a breaking change. Instead the library should just treat null & empty the same (NO_VERIFIER error).

Reply:
Ah, I see. That clears things up. 

Still, for the sync parser, treating them the same might break backwards compatibility. Right now, if the list of verifier is empty, the resulting error that will eventually be returned to the caller would be different (SignatureException vs IllegalStateException). For the async parser, we can still fix this behavior to be more clear by treating them the same in async provideVerifiers and/or indicating that the function should not return null (with a nonnull annotation).

Reply:
I agree that we can at least make the Async parser cleaner. I was worrying that if it returns null and the caller didn't treat the null, then there will be NPE.

Reply:
Yes, let's leave this one as-is and fix it in the async parser.